packages/theme/src/unified/UnifiedThemeProvider.tsx

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ReactNode } from 'react';
+import type { Theme as Mui4Theme } from '@material-ui/core/styles';
+import {
+  StyledEngineProvider,
+  Theme as Mui5Theme,
+  ThemeProvider as Mui5Provider,
+} from '@mui/material/styles';
+import {
+  StylesProvider as Mui5StylesProvider,
+  createGenerateClassName,
+} from '@mui/styles';
+import Mui5CssBaseline from '@mui/material/CssBaseline';
+import { UnifiedTheme } from './types';
+import { maybeLoadMui4CssBaseline, maybeLoadMui4Styles } from '../v4/load';
+
+const generateV4ClassName = createGenerateClassName({
+  seed: 'm4',
+});
+const generateV5ClassName = createGenerateClassName({
+  seed: 'm5',
+});
+
+/**
+ * Props for {@link UnifiedThemeProvider}.
+ *
+ * @public
+ */
+export interface UnifiedThemeProviderProps {
+  children: ReactNode;
+  theme: UnifiedTheme;
+  noCssBaseline?: boolean;
+}
+
+/**
+ * Provides themes for all MUI versions supported by the provided unified theme.
+ *
+ * @public
+ */
+export function UnifiedThemeProvider(
+  props: UnifiedThemeProviderProps,
+): JSX.Element {
+  const { children, theme, noCssBaseline = false } = props;
+
+  const v4Theme = theme.getTheme('v4');
+  const v5Theme = theme.getTheme('v5');
+
+  let cssBaseline: JSX.Element | undefined = undefined;
+  if (!noCssBaseline) {
+    if (v5Theme) {
+      cssBaseline = <Mui5CssBaseline enableColorScheme />;
+    } else if (v4Theme) {
+      const CssBaseline = maybeLoadMui4CssBaseline();
+      if (!CssBaseline) {
+        throw new Error('Failed to load MUI 4 CssBaseline component');
+      }
+      cssBaseline = <CssBaseline />;
+    }
+  }
+
+  let result = (
+    <>
+      {cssBaseline}
+      {children}
+    </>
+  );
+
+  if (v4Theme) {
+    const styles = maybeLoadMui4Styles();
+    if (!styles) {
+      throw new Error('Failed to load MUI 4 styles package');
+    }
+    const { StylesProvider, ThemeProvider } = styles;
+    result = (
+      <StylesProvider generateClassName={generateV4ClassName} injectFirst>
+        <ThemeProvider theme={v4Theme as Mui4Theme}>{result}</ThemeProvider>
+      </StylesProvider>
+    );
+  }
+
+  if (v5Theme) {
+    result = (
+      <Mui5StylesProvider generateClassName={generateV5ClassName}>
```

Comment:
This may not work as you would expect. For v5, we have introduced the [ClassNameGenerator](https://mui.com/material-ui/experimental-api/classname-generator/) that can do the same thing. Can you share some example of why it is needed? Is it part of the migration?

Reply:
It is marked as unstable so maybe that's it, or that it needs to be imported instead of passed as a prop for conditional naming? (at least according to the docs)

Reply:
Can it be configured and then passed here to the provider? if so then using v5 generator would be better

Reply:
@mnajdova I have a different read of this. Their use seems correct. The problem they seem to fix is the need for the JSS index counter to be different between @mui/styles and @material-ui/styles.

Reply:
> It is marked as unstable so maybe that's it, or that it needs to be imported instead of passed as a prop for conditional naming? (at least according to the docs)

Yes, it is unstable, but I don't expect changes in the API cc @siriwatknp 

> @mnajdova I have a different read of this. The problem they seem to fix is the need for the JSS index counter to be different between @mui/styles and @material-ui/styles.

If you are right then yes they don't need it, my thought is that it was required for: https://github.com/backstage/backstage/pull/15484/files#diff-77042c247fb20ea072a208a5e430bfaa6cb4b876837e130053bc12e24dd47023R66

If there is a way to avoid using it, it would be best, this is why I asked for some example of why they need it.

Reply:
Agree, I don't think the `generateClassName` is needed since v4 and v5 use different engine.